tests/dialects/test_riscv_snitch.py

```diff
@@ -0,0 +1,35 @@
+import pytest
+
+from xdsl.dialects import riscv_snitch
+from xdsl.dialects.riscv import RISCVInstruction
+from xdsl.traits import HasInsnRepresentation
+
+ground_truth = {
+    "dmsrc": ".insn r 0x2b, 0, 0, x0, {0}, {1}",
+    "dmdst": ".insn r 0x2b, 0, 1, x0, {0}, {1}",
+    "dmcpyi": ".insn r 0x2b, 0, 2, {0}, {1}, {2}",
+    "dmcpy": ".insn r 0x2b, 0, 3, {0}, {1}, {2}",
+    "dmstati": ".insn r 0x2b, 0, 4, {0}, {1}, {2}",
+    "dmstat": ".insn r 0x2b, 0, 5, {0}, {1}, {2}",
+    "dmstr": ".insn r 0x2b, 0, 6, x0, {0}, {1}",
+    "dmrep": ".insn r 0x2b, 0, 7, x0, {0}, x0",
+}
+
+
+@pytest.mark.parametrize(
+    "op",
+    (
+        riscv_snitch.DMSourceOp,
+        riscv_snitch.DMDestinationOp,
+        riscv_snitch.DMCopyImmOp,
+        riscv_snitch.DMCopyOp,
+        riscv_snitch.DMStatImmOp,
+        riscv_snitch.DMStatOp,
+        riscv_snitch.DMStrideOp,
+        riscv_snitch.DMRepOp,
+    ),
+)
+def test_insn_repr(op: RISCVInstruction):
+    trait = op.get_trait(HasInsnRepresentation)
+    assert trait is not None
+    assert trait.get_insn(op) == ground_truth[op.name[13:]]
```

Comment:
Pyright complains that: `Method "get_insn" cannot be called because it is abstract and unimplemented (reportGeneralTypeIssues)`

But `get_insn` is an abstract method on `HasInsnRepresentation`, which itself is ABC, so I can never get it "not implemented" as I will always get a subclass of this trait (that by definition has to implement the method). How the heck am I supposed to fix this? @superlopuh @math-fehr

Reply:
I think it's because the `get_insn` type parameter is invariant, meaning it says that it returns only instances of this type exactly, not subclasses. I think we want the covariant version instead.

Reply:
Turns out I was wrong, it was indeed a Pyright bug